modules/sfm/src/simple_pipeline.cpp

```diff
@@ -0,0 +1,444 @@
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of Willow Garage, Inc. nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include "precomp.hpp"
+
+#if CERES_FOUND
+
+#include <opencv2/xfeatures2d.hpp>
+#include <opencv2/sfm/simple_pipeline.hpp>
+
+#include "libmv/correspondence/feature.h"
+#include "libmv/correspondence/feature_matching.h"
+#include "libmv/correspondence/matches.h"
+#include "libmv/correspondence/nRobustViewMatching.h"
+
+#include "libmv/simple_pipeline/pipeline.h"
+#include "libmv/simple_pipeline/camera_intrinsics.h"
+#include "libmv/simple_pipeline/bundle.h"
+#include "libmv/simple_pipeline/initialize_reconstruction.h"
+#include "libmv/simple_pipeline/uncalibrated_reconstructor.h"
+#include "libmv/simple_pipeline/tracks.h"
+
+using namespace cv;
+using namespace std;
+
+namespace cv
+{
+namespace sfm
+{
+
+typedef struct libmv_ReconstructionBase
+{
+  /* used for per-track average error calculation after reconstruction */
+  libmv::Tracks tracks;
+  libmv::CameraIntrinsics intrinsics;
+
+  double error;
+
+} libmv_Reconstruction;
+
+
+typedef struct libmv_EuclideanReconstruction : public libmv_ReconstructionBase
+{
+  libmv::EuclideanReconstruction reconstruction;
+
+} libmv_EuclideanReconstruction;
+
+
+typedef struct libmv_ProjectiveReconstruction : public libmv_ReconstructionBase
+{
+  libmv::ProjectiveReconstruction reconstruction;
+
+} libmv_ProjectiveReconstruction;
+
+
+typedef struct libmv_UncalibratedReconstruction : public libmv_ReconstructionBase
+{
+  libmv::EuclideanReconstruction euclidean_reconstruction;
+  libmv::ProjectiveReconstruction projective_reconstruction;
+
+} libmv_UncalibratedReconstruction;
+
+
+void
+libmv_solveReconstruction( const libmv::Tracks &tracks,
```

Comment:
@sandwichmaker Any tip about how to improve the libmv pipeline? Sometimes depending on the initial reconstruction the results are really bad